File: backend/app/schemas/dossier.py
"""Pydantic v2 schemas for Dossier tracking (Phase 3)."""

from __future__ import annotations

import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import DossierStatut


# ── Dossier schemas ──────────────────────────────────────────────


class DossierCreate(BaseModel):
    """For programmatic import (not a public creation API)."""

    numero: str = Field(..., max_length=50)
    contact_id: uuid.UUID | None = None
    statut: DossierStatut = DossierStatut.en_attente
    type_projet: str | None = Field(None, max_length=200)
    raison_sociale: str | None = Field(None, max_length=300)
    montant_investissement: Decimal | None = Field(None, ge=0)
    region: str | None = Field(None, max_length=100)
    secteur: str | None = Field(None, max_length=200)
    date_depot: date | None = None
    date_derniere_maj: date | None = None
    observations: str | None = None
    raw_data: dict | None = None


class DossierRead(BaseModel):
    """Dossier response — list view (no history)."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    numero: str
    contact_id: uuid.UUID | None
    statut: DossierStatut
    type_projet: str | None
    raison_sociale: str | None
    montant_investissement: Decimal | None
    region: str | None
    secteur: str | None
    date_depot: date | None
    date_derniere_maj: date | None
    observations: str | None
    created_at: datetime
    updated_at: datetime


class DossierHistoryRead(BaseModel):
    """Single history entry for a dossier field change."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    field_changed: str
    old_value: str | None
    new_value: str | None
    changed_at: datetime
    sync_log_id: uuid.UUID | None


class DossierDetail(DossierRead):
    """Dossier detail response — includes full history."""

    history: list[DossierHistoryRead] = []


class DossierList(BaseModel):
    """Paginated list of dossiers."""

    items: list[DossierRead]
    total: int
    page: int
    page_size: int


class DossierStats(BaseModel):
    """Aggregated KPIs for back-office and Agent Interne."""

    total: int = 0
    en_cours: int = 0
    valide: int = 0
    rejete: int = 0
    en_attente: int = 0
    incomplet: int = 0


class DossierFilters(BaseModel):
    """Filters for paginated dossier listing."""

    statut: DossierStatut | None = None
    type_projet: str | None = None
    date_depot_from: date | None = None
    date_depot_to: date | None = None
    search: str | None = None  # search on numero, raison_sociale

File: frontend/src/types/dossier.ts
// ---------------------------------------------------------------------------
// Dossier types — mirrors backend DossierRead / DossierStats / Sync* schemas
// ---------------------------------------------------------------------------

import type { LucideIcon } from "lucide-react";
import {
  Clock,
  CheckCircle,
  XCircle,
  Hourglass,
  Paperclip,
  AlertTriangle,
  FolderOpen,
  Loader2,
  CheckCircle2,
  XOctagon,
} from "lucide-react";

// ---------------------------------------------------------------------------
// Enums
// ---------------------------------------------------------------------------

export type DossierStatut =
  | "en_cours"
  | "valide"
  | "rejete"
  | "en_attente"
  | "complement"
  | "incomplet";

export type SyncStatus = "pending" | "running" | "completed" | "failed";

export type SyncSourceType = "excel" | "csv" | "api_rest" | "manual";

export type SyncProviderType = "excel_csv" | "api_rest" | "db_link";

// ---------------------------------------------------------------------------
// Data interfaces
// ---------------------------------------------------------------------------

export interface Dossier {
  id: string;
  numero: string;
  contact_id: string | null;
  statut: DossierStatut;
  type_projet: string | null;
  raison_sociale: string | null;
  montant_investissement: string | null; // Decimal serialized as string
  region: string | null;
  secteur: string | null;
  date_depot: string | null;
  date_derniere_maj: string | null;
  observations: string | null;
  created_at: string;
  updated_at: string;
}

export interface DossierHistory {
  id: string;
  field_changed: string;
  old_value: string | null;
  new_value: string | null;
  changed_at: string;
  sync_log_id: string | null;
}

export interface DossierDetail extends Dossier {
  history: DossierHistory[];
}

export interface DossierStats {
  total: number;
  en_cours: number;
  valide: number;
  rejete: number;
  en_attente: number;
  complement: number;
  incomplet: number;
}

export interface SyncLog {
  id: string;
  source_type: SyncSourceType;
  file_name: string | null;
  file_hash: string | null;
  rows_total: number;
  rows_imported: number;
  rows_updated: number;
  rows_errored: number;
  error_details: Record<string, unknown> | null;
  status: SyncStatus;
  started_at: string | null;
  completed_at: string | null;
  triggered_by: string | null;
  created_at: string;
}

export interface SyncConfig {
  id: string;
  provider_type: SyncProviderType;
  config_json: Record<string, unknown>;
  column_mapping: Record<string, string>;
  schedule_cron: string | null;
  watched_folder: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface SyncConfigCreatePayload {
  provider_type?: SyncProviderType;
  config_json?: Record<string, unknown>;
  column_mapping: Record<string, string>;
  schedule_cron?: string;
  watched_folder?: string;
  is_active?: boolean;
}

// ---------------------------------------------------------------------------
// UI config — labels, styling, icons
// ---------------------------------------------------------------------------

export const STATUT_CONFIG: Record<
  DossierStatut,
  { label: string; className: string; icon: LucideIcon }
> = {
  en_cours: {
    label: "En cours",
    className: "bg-[#C4944B]/10 text-[#C4944B] border-[#C4944B]/20",
    icon: Clock,
  },
  valide: {
    label: "Validé",
    className: "bg-[#5F8B5F]/10 text-[#5F8B5F] border-[#5F8B5F]/20",
    icon: CheckCircle,
  },
  rejete: {
    label: "Rejeté",
    className: "bg-[#B5544B]/10 text-[#B5544B] border-[#B5544B]/20",
    icon: XCircle,
  },
  en_attente: {
    label: "En attente",
    className: "bg-[#5B7A8B]/10 text-[#5B7A8B] border-[#5B7A8B]/20",
    icon: Hourglass,
  },
  complement: {
    label: "Complément",
    className: "bg-[#C4704B]/10 text-[#C4704B] border-[#C4704B]/20",
    icon: Paperclip,
  },
  incomplet: {
    label: "Incomplet",
    className: "bg-muted text-muted-foreground border-muted-foreground/20",
    icon: AlertTriangle,
  },
};

/** KPI card config — total card + per-status cards */
export const STATS_CARD_CONFIG: Array<{
  key: keyof DossierStats;
  label: string;
  icon: LucideIcon;
  color: string; // Tailwind bg for icon circle
  textColor: string;
}> = [
  { key: "total", label: "Total dossiers", icon: FolderOpen, color: "bg-[#C4704B]/10", textColor: "text-[#C4704B]" },
  { key: "en_cours", label: "En cours", icon: Clock, color: "bg-[#C4944B]/10", textColor: "text-[#C4944B]" },
  { key: "valide", label: "Validés", icon: CheckCircle, color: "bg-[#5F8B5F]/10", textColor: "text-[#5F8B5F]" },
  { key: "rejete", label: "Rejetés", icon: XCircle, color: "bg-[#B5544B]/10", textColor: "text-[#B5544B]" },
  { key: "en_attente", label: "En attente", icon: Hourglass, color: "bg-[#5B7A8B]/10", textColor: "text-[#5B7A8B]" },
  { key: "complement", label: "Complément", icon: Paperclip, color: "bg-[#C4704B]/10", textColor: "text-[#C4704B]" },
  { key: "incomplet", label: "Incomplets", icon: AlertTriangle, color: "bg-muted", textColor: "text-muted-foreground" },
];

export const SYNC_STATUS_CONFIG: Record<
  SyncStatus,
  { label: string; className: string; icon: LucideIcon }
> = {
  pending: {
    label: "En attente",
    className: "bg-[#C4944B]/10 text-[#C4944B] border-0",
    icon: Clock,
  },
  running: {
    label: "En cours",
    className: "bg-[#5B7A8B]/10 text-[#5B7A8B] border-0",
    icon: Loader2,
  },
  completed: {
    label: "Terminé",
    className: "bg-[#5F8B5F]/10 text-[#5F8B5F] border-0",
    icon: CheckCircle2,
  },
  failed: {
    label: "Échoué",
    className: "bg-[#B5544B]/10 text-[#B5544B] border-0",
    icon: XOctagon,
  },
};

/** French labels for dossier fields (used in timeline) */
export const DOSSIER_FIELD_LABELS: Record<string, string> = {
  numero: "Numéro",
  statut: "Statut",
  type_projet: "Type de projet",
  raison_sociale: "Raison sociale",
  montant_investissement: "Montant investissement",
  region: "Région",
  secteur: "Secteur",
  date_depot: "Date de dépôt",
  date_derniere_maj: "Dernière mise à jour",
  observations: "Observations",
  contact_id: "Contact",
};

/** Target fields for column mapping in import wizard */
export const MAPPABLE_FIELDS: Array<{ value: string; label: string; required?: boolean }> = [
  { value: "numero", label: "Numéro de dossier", required: true },
  { value: "statut", label: "Statut" },
  { value: "type_projet", label: "Type de projet" },
  { value: "raison_sociale", label: "Raison sociale" },
  { value: "montant_investissement", label: "Montant investissement" },
  { value: "region", label: "Région" },
  { value: "secteur", label: "Secteur" },
  { value: "date_depot", label: "Date de dépôt" },
  { value: "date_derniere_maj", label: "Dernière mise à jour" },
  { value: "observations", label: "Observations" },
  { value: "phone", label: "Téléphone contact" },
];
